### ISSUE TITLE
[Question] Bypass A15 restrictions

### ISSUE THREAD

**Issue by pixincreate** [chenxiaolong/PixelLight#9]
Hi again!

I had been using PixelLight for a while and uninstalled it recently as I couldn't find a valid reason to it a default app until I README that said I can use it in QuickTile and also in the Lockscreen(!?).

But A15 has got a ton of limitations as compared to A14. So, the question is, if I make it a system app (by injecting it using `my-avbroot-setup`), do you think this can be bypassed? Asking so because I never really went through system level permissions that can bypass these restrictions.

I can truly add some `sed` commands to my repo for modifying `my-avbroot-setup`, or possibly raise a PR adding `PixelLight` support to `my-avbroot-setup`.

Thanks in advance!

**Comment by pixincreate** [comment]
For some reason, it always shows `Unavailable` in quick tile (A15)

**Comment by chenxiaolong** [comment]
> For some reason, it always shows `Unavailable` in quick tile (A15)

Hmm, could you grab a logcat immediately after expanding the quick settings panel once? That shouldn't happen once permissions have been granted.

> that said I can use it in QuickTile and also in the Lockscreen(!?).

Yep! With version 1.x, if you open the quick settings tiles while the device is still locked, the tile still works.

---

If features from 1.x are useful to you, I would recommend just installing that version. The last time I tested, Android 15 didn't enforce the new restrictions if the app targets API 34 (like PixelLight 1.x). I don't mention this in the README because I think it's a bug that this happens to work and is probably going to get patched in the future.

But to answer your question directly, Android seems to enforce most of its restrictions for both user apps and system apps. There's a chance that if PixelLight declared the `START_FOREGROUND_SERVICES_FROM_BACKGROUND` permission and it's also installed in `/system/priv-app/`, then it might be able to bypass the restrictions. I'm not sure though--it's a big pain tracing through [Android's logic](https://android.googlesource.com/platform/frameworks/base/+/refs/tags/android-15.0.0_r3/services/core/java/com/android/server/am/ActiveServices.java).

**Comment by chenxiaolong** [comment]
I also considered making PixelLight start a foreground service during boot that stays running forever and then the quick settings tile just connects to it. This approach is blocked by Android 15 too.

They're really trying to lock down things that automatically use the camera. If only Google allowed the maximum brightness via the flashlight API, then PixelLight wouldn't even need to exist...

**Comment by pixincreate** [comment]
> Hmm, could you grab a logcat immediately after expanding the quick settings panel once? That shouldn't happen once permissions have been granted.

Sure thing! I'm AFK at the moment, will get the logs logged in a separate issue.

> Yep! With version 1.x, if you open the quick settings tiles while the device is still locked, the tile still works.

Will give that a shot.

> But to answer your question directly, Android seems to enforce most of its restrictions for both user apps and system apps. There's a chance that if PixelLight declared the `START_FOREGROUND_SERVICES_FROM_BACKGROUND` permission and it's also installed in `/system/priv-app/`, then it might be able to bypass the restrictions. I'm not sure though--it's a big pain tracing through Android's logic.

Oh, putting this permission into the APP will clearly break the APP from running on rootless devices and it cannot be made optional either (if I'm not wrong).

Google's codebase is a rabbit hole unfortunately. Do you think cannot help here? I mean, even the ADB privileges do not have a bypass (?). I can take this as a homework BTW.

> They're really trying to lock down things that automatically use the camera. If only Google allowed the maximum brightness via the flashlight API, then PixelLight wouldn't even need to exist...

Google's really good at making sure that things are binded together -- Flashlight with Camera, Launcher with Recent menu, and the list goes on and on.

Off topic:

Now that I think about this, I'll give a shot to modify `my-avbroot-setup` to include pixel launcher and ASI by granting necessary permissiion, lol. Hopes are low (it might (will) cause bootloop).

**Comment by pixincreate** [comment]
For a moment, the quick tile button looked normal and I tried clicking on it. It crashed:

```
type: crash
osVersion: google/bluejay/bluejay:15/AP3A.241105.007/2024111800:user/release-keys
flags: dev options enabled
package: com.chiller3.pixellight:131072, targetSdk 35
process: com.chiller3.pixellight
processUptime: 9456 + 820 ms
installer: com.android.packageinstaller

java.lang.SecurityException: checkPidStatus:2206: Attempt to use camera from a different process than original client
	at android.hardware.camera2.utils.ExceptionUtils.throwAsPublicException(ExceptionUtils.java:68)
	at android.hardware.camera2.impl.ICameraDeviceUserWrapper.endConfigure(ICameraDeviceUserWrapper.java:115)
	at android.hardware.camera2.impl.CameraDeviceImpl.configureStreamsChecked(CameraDeviceImpl.java:609)
	at android.hardware.camera2.impl.CameraDeviceImpl.createCaptureSessionInternal(CameraDeviceImpl.java:824)
	at android.hardware.camera2.impl.CameraDeviceImpl.createCaptureSession(CameraDeviceImpl.java:770)
	at com.chiller3.pixellight.TorchSession$2$$ExternalSyntheticLambda0.run(Unknown Source:65)
	at android.os.Handler.handleCallback(Handler.java:959)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8744)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
	at com.android.internal.os.ExecInit.main(ExecInit.java:50)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:369)
Caused by: android.os.ServiceSpecificException: checkPidStatus:2206: Attempt to use camera from a different process than original client (code 1)
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3265)
	at android.os.Parcel.createException(Parcel.java:3235)
	at android.os.Parcel.readException(Parcel.java:3211)
	at android.os.Parcel.readException(Parcel.java:3153)
	at android.hardware.camera2.ICameraDeviceUser$Stub$Proxy.endConfigure(ICameraDeviceUser.java:703)
	at android.hardware.camera2.impl.ICameraDeviceUserWrapper.endConfigure(ICameraDeviceUserWrapper.java:112)
	... 14 more
```

**Comment by chenxiaolong** [comment]
Wow, that is incredibly weird. Any chance you can grab a full logcat when that happens?

From looking at the AOSP code, Android should be logging a message like:
```
attempt to use a locked camera from a different process (old pid %d, new pid %d)
```

If you see that message, can you check `ps -ef | grep PID` to see what the 2 processes are?

I'd expect one of them to be PixelLight (or more likely the PID is missing since it's crashing), but I have no idea what the other one could be. Maybe something else is trying to access the camera at the same time?

**Comment by chenxiaolong** [comment]
> even the ADB privileges do not have a bypass (?).

Unfortunately not (at least not that I can see)

---

One thing that I plan on trying is to (optionally) leave the foreground service running forever after the user turns on the flashlight once. And then further uses of the quick settings tile would just connect to the existing foreground service and reuse it.

If this works, the first use of the tile after a reboot will behave like version 2.x (doesn't work on lock screen and causes the panel to close), but further uses behave like version 1.x.

I don't think this should impact battery life (the service would be completely idle) and the user can still disable the (mandatory) notification from Android's settings if it's annoying.

**Comment by pixincreate** [comment]
> Wow, that is incredibly weird. Any chance you can grab a full logcat when that happens?
> 
> From looking at the AOSP code, Android should be logging a message like:
> 
> ```
> attempt to use a locked camera from a different process (old pid %d, new pid %d)
> ```
> 
> If you see that message, can you check `ps -ef | grep PID` to see what the 2 processes are?
> 
> I'd expect one of them to be PixelLight (or more likely the PID is missing since it's crashing), but I have no idea what the other one could be. Maybe something else is trying to access the camera at the same time?

Will do it today.



> > even the ADB privileges do not have a bypass (?).
> 
> Unfortunately not (at least not that I can see)
> 
> One thing that I plan on trying is to (optionally) leave the foreground service running forever after the user turns on the flashlight once. And then further uses of the quick settings tile would just connect to the existing foreground service and reuse it.
> 
> If this works, the first use of the tile after a reboot will behave like version 2.x (doesn't work on lock screen and causes the panel to close), but further uses behave like version 1.x.
> 
> I don't think this should impact battery life (the service would be completely idle) and the user can still disable the (mandatory) notification from Android's settings if it's annoying.

This does sounds good. But the user will be forced to have an app running in background actively.

**Comment by chenxiaolong** [comment]
The new approach works! (#10) The user experience is not perfect of course, but still better than how version 2.0 worked.

**Comment by e-t-l** [comment]
I'm a little confused, what's the benefit of using v2.1 over v1.0? I'm on Android 15, installed v1.0, and was able to turn on the flashlight from the PixelLight tile from the lockscreen. So if A15 restricts the v2.x, why bother updating the app? (at least until Android presumably patches out its functionality)

**Comment by chenxiaolong** [comment]
I'd suggest just sticking with v1.0. For now, Google doesn't enforce the new restrictions for apps targeting Android 14 (like v1.0). I'd expect this to change at any time because v1.0 uses APIs in an unusual way that Google does not guarantee backwards compatibility for. When it finally breaks, you can update to 2.x, which doesn't use any weird hacks to make the functionality work.

**Comment by chenxiaolong** [comment]
Closing this issue due to #12/#13. PixelLight 2.2 can now always turn on the flashlight while the device is locked. There's still unfixable UI jank due to Android animation bugs, but there's no longer any regression in functionality compared to 1.0.
